man/im_rightshift_size.3
.TH IM_RIGHTSHIFT_SIZE 3 "12 August 2006"
.SH NAME
 im_rightshift_size \- Decrease size by a power-of-two factor
.SH SYNOPSIS
.nf
.B #include <vips/vips.h>
.sp
.BI "int im_rightshift_size( IMAGE " "*in" ", IMAGE " "*out" ", int " "xshift" ", int " "yshift" ", int " "band_fmt" " );
.fi
.SH DESCRIPTION
.B im_rightshift_size(3)
decreases the size of an integer type image by a power-of-two factor, very quickly, by summing the values of
adjacent pixels.  The sum is shifted to produce output of the specified band format.  
.IR xshift " and "yshift
are positive and give the base-two logarithms of the scale factors.
.PP
If the input image is a signed type, then 
.I band_fmt
must be one of IM_BANDFMT_CHAR, IM_BANDFMT_SHORT or IM_BANDFMT_INT.  If it is an unsigned type, then
.I band_fmt
must be one of IM_BANDFMT_UCHAR, IM_BANDFMT_USHORT or IM_BANDFMT_UINT.
.SH RETURN VALUE
The functions returns 0 on success and -1 on error.
.SH SEE ALSO
im_affine(3)
.SH COPYRIGHT
.br
Copyright 2006, Tom Vajzovic.
.SH AUTHOR
Tom Vajzovic


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

libsrc/conversion/im_rightshift_size.c
/* @(#) Decrease the size of an image by a power-of-two factor, summing the
 * @(#) values of pixels, and shifting to give output of the specified band
 * @(#) format.
 * @(#)
 * @(#) int
 * @(#) im_rightshift_size(
 * @(#)   IMAGE *in,
 * @(#)   IMAGE *out,
 * @(#)   int xshift,
 * @(#)   int yshift,
 * @(#)   int band_fmt
 * @(#) );
 * @(#)
 *
 * Copyright: 2006, Tom Vajzovic
 *
 * Author: Tom Vajzovic
 *
 * Written on: 2006-07-26
 *
 * 2006-12-17 tcv: 
 *  - rewrite generator function macros to produce many smaller functions
 *  - remove post-shift-up functionality
 * 2007-02-02 jc
 *  - added return 0; on success
 *  - FUNCTION_NAME updated
 */

/*

    This file is part of VIPS.

    VIPS is free software; you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

 */

/*

    These files are distributed with VIPS - http://www.vips.ecs.soton.ac.uk

 */


/** HEADERS **/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif /*HAVE_CONFIG_H */
#include <vips/intl.h>

#include <stdlib.h>
#include <string.h>

#include <vips/vips.h>

#ifdef WITH_DMALLOC
#include <dmalloc.h>
#endif /*WITH_DMALLOC */


/** LOCAL FUNCTION DECLARATIONS **/

/* function prototype macro */
#define GEN_FUNC( SHIFT_MACRO, FROM_T, TO_T, SUM_T ) \
static int gen_ ## SHIFT_MACRO ## _ ## FROM_T ## _to_ ## TO_T ## _with_ ## SUM_T( \
          REGION *to_make, REGION *make_from, void *unrequired, int *params );

/* macros to call function prototype macro for all possible combinations of types and macros */
#define GEN_FUNCS_TO( SIGN, FROM_T, TO_T )    \
  GEN_FUNC( PRE_POST_SHIFT, FROM_T, TO_T, SIGN ## 64 ) \
  GEN_FUNC( POST_SHIFT, FROM_T, TO_T, SIGN ## 64 ) \
  GEN_FUNC( POST_SHIFT, FROM_T, TO_T, SIGN ## 32 ) \
  GEN_FUNC( NO_SHIFT, FROM_T, TO_T, SIGN ## 32 )

#define GEN_FUNCS_FROM( SIGN, FROM_T )      \
  GEN_FUNCS_TO( SIGN, FROM_T, SIGN ## 32 )  \
  GEN_FUNCS_TO( SIGN, FROM_T, SIGN ## 16 )  \
  GEN_FUNCS_TO( SIGN, FROM_T, SIGN ## 8 )

#define GEN_FUNCS_SIGN( SIGN )          \
  GEN_FUNCS_FROM( SIGN, SIGN ## 32 )    \
  GEN_FUNCS_FROM( SIGN, SIGN ## 16 )    \
  GEN_FUNCS_FROM( SIGN, SIGN ## 8 )

GEN_FUNCS_SIGN( gint )
GEN_FUNCS_SIGN( guint )


/** FUNCTION DEFINITIONS **/

int
im_rightshift_size( IMAGE *in, IMAGE *out, int xshift, int yshift, int band_fmt ){
#define FUNCTION_NAME "im_rightshift_size"
  int *params;
  int needbits;
  int outbits;

  if( im_piocheck( in, out ) )
    return -1;

  if( xshift < 0 || yshift < 0 ){
    im_error( FUNCTION_NAME, _( "bad arguments" ) );
    return -1;
  }
  if( ! xshift && ! yshift ){
    im_warn( FUNCTION_NAME, _( "shift by zero: falling back to im_copy" ) );
    return im_copy( in, out );
  }
  if( ! in-> Xsize >> xshift || ! in-> Ysize >> yshift ){
    im_error( FUNCTION_NAME, _( "would result in zero size output image" ) );
    return -1;
  }
  if( ! im_isint( in ) ){
    im_error( FUNCTION_NAME, _( "integer type images only" ) );
    return -1;
  }
  if( IM_CODING_NONE != in->Coding ){
    im_error( FUNCTION_NAME, _( "uncoded images only" ) );
    return -1;
  }
  if( im_isuint( in ) ){
    if( IM_BANDFMT_UCHAR != band_fmt && IM_BANDFMT_USHORT != band_fmt && IM_BANDFMT_UINT != band_fmt ){
      im_error( FUNCTION_NAME, _( "unsigned input means that output must be unsigned int, short or char" ) );
      return -1;
    }
  }
  else {
    if( IM_BANDFMT_CHAR != band_fmt && IM_BANDFMT_SHORT != band_fmt && IM_BANDFMT_INT != band_fmt ){
      im_error( FUNCTION_NAME, _( "signed input means that output must be signed int, short or char" ) );
      return -1;
    }
  }
  outbits= im_bits_of_fmt( band_fmt );

  if( -1 == outbits )
    return -1;

  needbits= im_bits_of_fmt( in-> BandFmt ) + xshift + yshift;

  params= IM_ARRAY( out, 4, int );

  if( ! params )
    return -1;

  params[ 0 ]= xshift;                      /* xshift */
  params[ 1 ]= yshift;                      /* yshift */

  if( im_cp_desc( out, in ) )
    return -1;

  out-> Xsize >>= xshift;
  out-> Ysize >>= yshift;
  out-> Xres/= ( 1 << xshift ); /* x scale */
  out-> Yres/= ( 1 << yshift ); /* y scale */
  out-> BandFmt= band_fmt;
  out-> Bbits= outbits;

  if( im_demand_hint( out, IM_THINSTRIP, in, NULL ) )
    return -1;

#define RETURN_MACRO( MACRO, FROM_T, TO_T, SUM_T ) \
  return im_generate( out, im_start_one, gen_ ## MACRO ## _ ## FROM_T ## _to_ ## TO_T ## _with_ ## SUM_T, im_stop_one, in, params );

#define RETURN_MACRO_OUTS( MACRO, SUM_SIZE, OUT_SIZE )   switch( in-> BandFmt ){  \
                                                                                  \
  case IM_BANDFMT_CHAR:                                                           \
    RETURN_MACRO( MACRO, gint8, gint ## OUT_SIZE, gint ## SUM_SIZE )              \
                                                                                  \
  case IM_BANDFMT_UCHAR:                                                          \
    RETURN_MACRO( MACRO, guint8, guint ## OUT_SIZE, guint ## SUM_SIZE )           \
                                                                                  \
  case IM_BANDFMT_SHORT:                                                          \
    RETURN_MACRO( MACRO, gint16, gint ## OUT_SIZE, gint ## SUM_SIZE )             \
                                                                                  \
  case IM_BANDFMT_USHORT:                                                         \
    RETURN_MACRO( MACRO, guint16, guint ## OUT_SIZE, guint ## SUM_SIZE )          \
                                                                                  \
  case IM_BANDFMT_INT:                                                            \
    RETURN_MACRO( MACRO, gint32, gint ## OUT_SIZE, gint ## SUM_SIZE )             \
                                                                                  \
  case IM_BANDFMT_UINT:                                                           \
    RETURN_MACRO( MACRO, guint32, guint ## OUT_SIZE, guint ## SUM_SIZE )          \
}

#define RETURN_MACRO_SUMSS( MACRO, SUM_SIZE )   switch( out-> Bbits ){  \
  case 32:                                                              \
    RETURN_MACRO_OUTS( MACRO, SUM_SIZE, 32 )                            \
  case 16:                                                              \
    RETURN_MACRO_OUTS( MACRO, SUM_SIZE, 16 )                            \
  case 8:                                                               \
    RETURN_MACRO_OUTS( MACRO, SUM_SIZE, 8 )                             \
}

  if( needbits > 64 ){
    params[ 2 ]= needbits - 64;
    params[ 3 ]= 64 - outbits;
    RETURN_MACRO_SUMSS( PRE_POST_SHIFT, 64 )
  }
  if( needbits > 32 ){
    params[ 3 ]= needbits - outbits;
    RETURN_MACRO_SUMSS( POST_SHIFT, 64 )
  }
  if( needbits > outbits ){
    params[ 3 ]= needbits - outbits;
    RETURN_MACRO_SUMSS( POST_SHIFT, 32 )
  }

  RETURN_MACRO_SUMSS( NO_SHIFT, 32 )

  return 0;

#undef FUNCTION_NAME
}

/** LOCAL FUNCTION DEFINITIONS **/

/* macros used in the function creating macro */
#define PRE_POST_SHIFT()    pixel+= from[ fx ] >> preshift; \
                          to[ tx ]= pixel >> postshift;

#define POST_SHIFT()   pixel+= from[ fx ];                  \
                          to[ tx ]= pixel >> postshift;

#define NO_SHIFT()          pixel+= from[ fx ];             \
                          to[ tx ]= pixel;

/* function creating macro */
#undef GEN_FUNC
#define GEN_FUNC( SHIFT_MACRO, FROM_T, TO_T, SUM_T ) \
static int gen_ ## SHIFT_MACRO ## _ ## FROM_T ## _to_ ## TO_T ## _with_ ## SUM_T(                 \
          REGION *to_make, REGION *make_from, void *unrequired, int *params ){                    \
                                                                                                  \
  int xshift= params[0];                                                                          \
  int yshift= params[1];                                                                          \
  int preshift= params[2];                                                                        \
  int postshift= params[3];                                                                       \
  Rect need= {                                                                                    \
    to_make-> valid. left << xshift,                                                              \
    to_make-> valid. top << yshift,                                                               \
    to_make-> valid. width << xshift,                                                             \
    to_make-> valid. height << yshift                                                             \
  };                                                                                              \
  TO_T *to_start= (TO_T*) IM_REGION_ADDR( to_make, to_make-> valid. left, to_make-> valid. top ); \
  TO_T *to, *ty_stop;                                                                             \
  SUM_T pixel;                                                                                    \
  FROM_T *from_start, *from_row, *from, *fy_stop;                                                 \
                                                                                                  \
  int band, tx, fx;                                                                               \
  int tx_max= to_make-> valid. width * to_make-> im-> Bands;                                      \
  int fx_max= to_make-> im-> Bands << xshift;                                                     \
                                                                                                  \
  size_t t_skip= IM_REGION_LSKIP( to_make ) / sizeof( TO_T );                                     \
  size_t f_skip;                                                                                  \
                                                                                                  \
  if( im_prepare( make_from, &need ) || ! im_rect_includesrect( &make_from-> valid, &need ) )     \
    return -1;                                                                                    \
                                                                                                  \
  from_start= (FROM_T*) IM_REGION_ADDR( make_from, need. left, need. top );                       \
  f_skip= IM_REGION_LSKIP( make_from ) / sizeof( FROM_T );                                        \
                                                                                                  \
  for( band= 0; band < make_from-> im-> Bands; ++band ){                                          \
                                                                                                  \
    to= to_start + band;                                                                          \
    ty_stop= to + t_skip * to_make-> valid. height;                                               \
                                                                                                  \
    from_row= from_start + band;                                                                  \
                                                                                                  \
    for( ; to < ty_stop; to+= t_skip, from_row+= f_skip << yshift )                               \
      for( tx= 0; tx < tx_max; tx+= to_make-> im-> Bands ){                                       \
                                                                                                  \
        from= from_row + ( tx << xshift );                                                        \
        fy_stop= from + ( f_skip << yshift );                                                     \
                                                                                                  \
        pixel= 0;                                                                                 \
                                                                                                  \
        for( ; from < fy_stop; from+= f_skip )                                                    \
          for( fx= 0; fx < fx_max; fx+= to_make-> im-> Bands )                                    \
            SHIFT_MACRO()                                                                         \
      }                                                                                           \
  }                                                                                               \
  return 0;                                                                                       \
}

/* create functions for all possible combinations of types and macros */
GEN_FUNCS_SIGN( gint )
GEN_FUNCS_SIGN( guint )
